sdks/java/core/src/main/java/org/apache/beam/sdk/values/Row.java

```diff
@@ -554,6 +555,12 @@ public Builder withFieldValueGetters(
       return this;
     }
 
+    /** The FieldValueGetters will handle the conversion for Arrays, Maps and Rows. */
+    public Builder withFieldValueGettersHandleCollections(boolean collectionHandledByGetter) {
+      this.collectionHandledByGetter = collectionHandledByGetter;
+      return this;
+    }
```

Comment:
OK, maybe the name could be better, but it means that the the FieldValueGetters also handle collections like ARRAY, ROW, MAP. If you go to the original implementation ( https://github.com/apache/beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/values/RowWithGetters.java#L90 ) you see that RowWithGetters does naive handling of ARRAY, ROW and MAP, for protobuf you need more context (the descriptor) to handle them. That's why I need to disable the naive mapping and let the FieldValueGetters handle ARRAY, ROW and MAP.

Feel free to suggest a better name though.

Reply:
@reuvenlax is this the last remaining concern you have for this PR?

https://github.com/apache/beam/pull/8690#issuecomment-497695744 is a good reference for the motivation of this as well.

I'm not crazy about the way this is implemented since it's adding state to `RowWithGetter` that will get checked every time a collection field is accessed. I can't think of a better way to do it without some non-trivial refactoring though. Some ideas:

1.  Rather than storing `collectionHandledByGetter` in `RowWithGetters` and changing behavior based on it, have two alternate `RowWithGetters` implementations, one with the special handling for collections and one without. I think this is still an improvement over the separate `ProtoRow` class that was rejected since it's not explicitly tied to a particular `SchemaProvider`, in fact it sounds like it could be re-used for Avro `GenericRecord` instances.
1. Move the  special logic for collection getters into those `SchemaProvider`/`FieldValueGetter` implementations that need it, and make `RowWithGetters` always behave as if `collectionHandledByGetter` is true. I think this could be a cleaner approach? But it's challenging because that special logic stores cached values that wouldn't be appropriate to move into the `FieldValueGetter` implementations. Maybe a `SchemaProvider` could have some way to indicate that a certain row is expensive to access and should be cached in `RowWithGetters`?

Reply:
I've refactored from your input: I've created a RowWithGettersCachedCollection that inherits for RowWithGetters. This cached is the default.

Reply:
Can you help me understand this a bit more? Why does it not work to cache lists for protocol buffers? We saw repeated array conversion to be a big problem (which is why we cache them). I'm wondering if we could instead cache a lazy array like we do with iterables.

I'll take a closer look at this code to figure it out.

Reply:
Would we make a ticket out of this or is this blocking?